{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fine-Mapping Preprocessing\n",
    "\n",
    "Jupyter Noteboook Tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup OpenAI API endpoint for ChatGPT integration\n",
    "\n",
    "### Run the following commands in your terminal:\n",
    "\n",
    "\n",
    "`$ export OPENAI_ORGANIZATION='<KEY>'`        # https://platform.openai.com/settings/organization/general\n",
    "\n",
    "`$ export OPENAI_PROJECT='<KEY>'`             # https://platform.openai.com/settings/ -> project\n",
    "\n",
    "`$ export OPENAI_API_KEY='<KEY>'`             # https://platform.openai.com/settings/profile?tab=api-keys\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "\n",
    "# organization = os.getenv('OPENAI_ORGANIZATION')\n",
    "# project = os.getenv('OPENAI_PROJECT')\n",
    "# api_key = os.getenv('OPENAI_API_KEY')\n",
    "\n",
    "# these are Sam's personal keys -- you can use until it runs out (caps out at $5 usage total)\n",
    "organization = 'org-7Tz4ejgn3Rw1nC3LMSjTcdVN'\n",
    "project = 'proj_JwOi33NfPxndlxwTCQL9UIiK'\n",
    "api_key = 'sk-None-DvEi4YfrZBudVnrZK1nTT3BlbkFJQsggOLv29hV8bq6Sm8Mf'\n",
    "\n",
    "openai_client = OpenAI(\n",
    "    organization = organization,\n",
    "    project = project,\n",
    "    api_key = api_key\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use path to folder containing your various sumstats files\n",
    "\n",
    "Verify these paths are correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90000583_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007228_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092856_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086092_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007545_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006906_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90014122_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092916_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092844_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST010774_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092831_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043954_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086091_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092904_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90025953_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90014243_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086077_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90245878_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269905_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104543_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90026612_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006910_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093050_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092892_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086137_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093040_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90000582_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009541_buildGRCh37.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009248_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043957_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104541_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90239676_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092930_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009249_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093016_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092868_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269497_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006907_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90027266_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269616_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90274723_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST011365_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093004_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269904_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092944_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009250_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086134_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90271714_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092985_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269498_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90013537_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093028_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092956_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90310293_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092968_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90239664_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90132315_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043977_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104542_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90271713_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007546_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009602_buildGRCh37.txt\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006909_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST004787_buildGRCh37.txt.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90301954_buildGRCh37.tsv\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90044025_buildGRCh37.tsv.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006414_buildGRCh37.tbl.gz\n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90132314_buildGRCh37.tsv\n"
     ]
    }
   ],
   "source": [
    "directory_of_sumstats = '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European' # YOUR INPUT PATH\n",
    "my_input_folder = glob.glob(directory_of_sumstats + '/*')\n",
    "\n",
    "for path in my_input_folder:\n",
    "    print(path)\n",
    "\n",
    "output_directory = '' # YOUR OUTPUT PATH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Instantiate Preprocess class with your \n",
    "    - OpenAI client\n",
    "    - Directory where you want output files located"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from preprocessing import Preprocess\n",
    "\n",
    "ft = Preprocess(\n",
    "    client=openai_client, \n",
    "    out_dir=output_directory\n",
    ")\n",
    "\n",
    "# optional parameters\n",
    "ft.significance_threshold = 5e-8 # default value is already 5e-8\n",
    "ft.ancestry = 'EUR' # default value is 'EUR' but can be statically\n",
    "# changed here or dynamically changed in the loop below.\n",
    "# Ancestry may play a role in GPT's selection of a column, e.g., say there\n",
    "# are two pval columns (1) pval_afr (2) pval_eur, then GPT will select\n",
    "# the column matching the ancestry variable set by user."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Iterate over each file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "68"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(my_input_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run one example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104541_buildGRCh37.tsv.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found GCST90104541_buildGRCh37.tsv.gz\n",
      "\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value', 'odds_ratio', 'ci_lower', 'ci_upper', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "\n",
      "Saved chunk to \n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "Saved chunk to \n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "Saved chunk to \n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "Saved chunk to \n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "Saved reformatted sumstats file to: GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "res = ft.loadmap_sumstats_table(\n",
    "    fp,\n",
    "    verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expects pre-filtered sumstats\n",
      "  CHR         BP         locus\n",
      "0   1  170193825   1.170193825\n",
      "1   4  111688752   4.111688752\n",
      "2   4  187213883   4.187213883\n",
      "3   9  136150466   9.136150466\n",
      "4  10  105324774  10.105324774\n",
      "\n",
      "Saved lead SNP file to \n",
      "/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed_leadSNPs.tsv\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>170193825</td>\n",
       "      <td>1.170193825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>111688752</td>\n",
       "      <td>4.111688752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>187213883</td>\n",
       "      <td>4.187213883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>136150466</td>\n",
       "      <td>9.136150466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>105324774</td>\n",
       "      <td>10.105324774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>16</td>\n",
       "      <td>73048367</td>\n",
       "      <td>16.73048367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHR         BP         locus\n",
       "0   1  170193825   1.170193825\n",
       "1   4  111688752   4.111688752\n",
       "2   4  187213883   4.187213883\n",
       "3   9  136150466   9.136150466\n",
       "4  10  105324774  10.105324774\n",
       "5  16   73048367   16.73048367"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ft.create_leadsnp_table(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run a batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90000583_buildGRCh37.tsv 0\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'p_value', 'variant_id', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'NA', 'NA', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007228_buildGRCh37.tsv 0\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'p_value', 'chr:pos(hg19)', 'effect_allele', 'other_allele', 'eaf', 'OR', 'OR_se', 'OR_95L', 'OR_95U', 'z', 'p.value', 'X_.log10_p.value', 'q_statistic', 'q_p.value', 'i2', 'n_studies', 'n_samples', 'effects'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'NA', 'OR_se', 'p_value', 'OR']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092856_buildGRCh37.tsv.gz 0\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086092_buildGRCh37.tsv 0\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rs_id_all', 'info_all', 'other_allele', 'effect_allele', 'all_maf', 'all_total', 'standard_error', 'beta', 'p_value', 'freq_effect_allele', 'chr_X_freq_alleleB_all', 'chr_X_freq_alleleB_males', 'chr_X_freq_alleleB_females', 'chr_X_all_total', 'chr_X_frequentist_add_pvalue_males', 'chr_X_frequentist_add_beta_1.genotype.sex.1_males', 'chr_X_frequentist_add_se_1.genotype.sex.1_males', 'chr_X_all_total_males', 'chr_X_frequentist_add_pvalue_females', 'chr_X_frequentist_add_beta_1.genotype.sex.2_females', 'chr_X_frequentist_add_se_1.genotype.sex.2_females', 'chr_X_all_total_females'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90086092_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007545_buildGRCh37.txt 1\n",
      "Sumstat cols:\t['SNP Position EA OA beta.CAD beta.lipid eaf.CAD eaf.lipid se.lipid pval.lipid se.CAD pval.CAD P'] ->\n",
      "GPT mapping:\t['NA', 'Position', 'EA', 'OA', 'beta.CAD', 'se.CAD', 'pval.CAD', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006906_buildGRCh37.tsv 1\n",
      "Sumstat cols:\t['MarkerName Allele1 Allele2 Freq1 Effect StdErr P-value'] ->\n",
      "GPT mapping:\t['MarkerName', 'NA', 'Allele1', 'Allele2', 'Effect', 'StdErr', 'P-value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90014122_buildGRCh37.tsv 1\n",
      "Sumstat cols:\t['variant_id', 'chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NStudies'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90014122_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092916_buildGRCh37.tsv.gz 2\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092844_buildGRCh37.tsv.gz 2\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST010774_buildGRCh37.tsv 2\n",
      "Sumstat cols:\t['variant_id', 'MAF_calculated_from_dosage_data', 'chromosome', 'base_pair_location', 'REF_allele', 'ALT_allele', 'p_value', 'p_value_SPACox_NoSPA', 'score_statistics_of_SPACox', 'empirical_variance', 'z_value', 'odds_ratio', 'ci_lower', 'ci_upper', 'beta', 'standard_error', 'effect_allele_frequency', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'ALT_allele', 'REF_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST010774_buildGRCh37_preprocessed.tsv\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092831_buildGRCh37.tsv.gz 3\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value', 'chromosome', 'base_pair_location'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90092831_buildGRCh37_preprocessed.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/groups/sanjana_lab/sfriedman/finemapping/FiMTools/src/preprocessing_v04.py:354: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for chunk in pd.read_table(\n",
      "/gpfs/commons/groups/sanjana_lab/sfriedman/finemapping/FiMTools/src/preprocessing_v04.py:364: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  sumstats_df = pd.concat(chunks)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043954_buildGRCh37.tsv.gz 4\n",
      "Sumstat cols:\t['chromosome', 'variant_id', 'base_pair_location', 'effect_allele', 'other_allele', 'N', 'effect_allele_frequency', 'T', 'SE_T', 'P_noSPA', 'beta', 'standard_error', 'p_value', 'CONVERGE'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90043954_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086091_buildGRCh37.tsv 5\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rs_id_all', 'info_all', 'other_allele', 'effect_allele', 'all_maf', 'all_total', 'freq_effect_allele', 'standard_error', 'beta', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90086091_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092904_buildGRCh37.tsv.gz 6\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90025953_buildGRCh37.tsv 6\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'GENPOS', 'ALLELE1', 'ALLELE0', 'A1FREQ', 'INFO', 'CHISQ_LINREG', 'P_LINREG', 'beta', 'standard_error', 'CHISQ_BOLT_LMM_INF', 'P_BOLT_LMM_INF', 'CHISQ_BOLT_LMM', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'ALLELE1', 'ALLELE0', 'beta', 'standard_error', 'P_BOLT_LMM_INF', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90025953_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90014243_buildGRCh37.tsv 7\n",
      "Sumstat cols:\t['chromosome', 'variant_id', 'base_pair_location', 'effect_allele', 'F_A', 'effect_allele_frequency', 'other_allele', 'CHISQ', 'p_value', 'odds_ratio', 'standard_error', 'ci_lower', 'ci_upper', 'BONF', 'FDR_BH'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'NA', 'standard_error', 'p_value', 'odds_ratio']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086077_buildGRCh37.tsv 7\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rs_id_all', 'info_all', 'other_allele', 'effect_allele', 'all_maf', 'all_total', 'standard_error', 'beta', 'p_value', 'freq_effect_allele', 'chr_X_freq_alleleB_all', 'chr_X_freq_alleleB_males', 'chr_X_freq_alleleB_females', 'chr_X_all_total', 'chr_X_frequentist_add_pvalue_males', 'chr_X_frequentist_add_beta_1.genotype.sex.1_males', 'chr_X_frequentist_add_se_1.genotype.sex.1_males', 'chr_X_all_total_males', 'chr_X_frequentist_add_pvalue_females', 'chr_X_frequentist_add_beta_1.genotype.sex.2_females', 'chr_X_frequentist_add_se_1.genotype.sex.2_females', 'chr_X_all_total_females'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90086077_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90245878_buildGRCh37.tsv.gz 8\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rsID', 'EA', 'OA', 'Ncases', 'Nctrls', 'EAF', 'BETA', 'SE', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'EA', 'OA', 'BETA', 'SE', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90245878_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269905_buildGRCh37.tsv.gz 9\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90269905_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104543_buildGRCh37.tsv.gz 10\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value', 'odds_ratio', 'ci_lower', 'ci_upper', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST90104543_buildGRCh37_preprocessed.tsv\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90026612_buildGRCh37.tsv 11\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'SNP', 'OA', 'EA', 'EAF', 'BETA', 'SE', 'p_value', 'Het_P', 'N_cases', 'N_ctrls'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'EA', 'OA', 'BETA', 'SE', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90026612_buildGRCh37_preprocessed.tsv\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006910_buildGRCh37.tsv 12\n",
      "Sumstat cols:\t['MarkerName Allele1 Allele2 Freq1 Effect StdErr P-value'] ->\n",
      "GPT mapping:\t['NA', 'NA', 'Allele1', 'Allele2', 'Effect', 'StdErr', 'P-value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093050_buildGRCh37.tsv.gz 12\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092892_buildGRCh37.tsv.gz 12\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086137_buildGRCh37.tsv 12\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rs_id_all', 'info_all', 'other_allele', 'effect_allele', 'all_maf', 'all_total', 'standard_error', 'beta', 'p_value', 'freq_effect_allele', 'chr_X_freq_alleleB_all', 'chr_X_freq_alleleB_males', 'chr_X_freq_alleleB_females', 'chr_X_all_total', 'chr_X_frequentist_add_pvalue_males', 'chr_X_frequentist_add_beta_1.genotype.sex.1_males', 'chr_X_frequentist_add_se_1.genotype.sex.1_males', 'chr_X_all_total_males', 'chr_X_frequentist_add_pvalue_females', 'chr_X_frequentist_add_beta_1.genotype.sex.2_females', 'chr_X_frequentist_add_se_1.genotype.sex.2_females', 'chr_X_all_total_females'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90086137_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093040_buildGRCh37.tsv.gz 13\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90000582_buildGRCh37.tsv 13\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'p_value', 'variant_id', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'NA', 'NA', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009541_buildGRCh37.gz 13\n",
      "Only supports dict_keys(['.tsv', '.csv', '.txt', '.tbl']) (+ .gz)\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009248_buildGRCh37.txt 13\n",
      "Sumstat cols:\t['Variant,CHR,BP,ALLELE1,ALLELE0,A1FREQ,INFO,BETA,SE,P,id'] ->\n",
      "GPT mapping:\t['CHR', 'BP', 'ALLELE1', 'ALLELE0', 'BETA', 'SE', 'P', 'NA']\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043957_buildGRCh37.tsv.gz 13\n",
      "Sumstat cols:\t['chromosome', 'variant_id', 'base_pair_location', 'effect_allele', 'other_allele', 'N', 'effect_allele_frequency', 'T', 'SE_T', 'P_noSPA', 'beta', 'standard_error', 'p_value', 'CONVERGE'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90043957_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104541_buildGRCh37.tsv.gz 14\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value', 'odds_ratio', 'ci_lower', 'ci_upper', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST90104541_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90239676_buildGRCh37.tsv 15\n",
      "Sumstat cols:\t['variant_id', 'chromosome', 'base_pair_location', 'other_allele', 'effect_allele', 'n', 'N_studies', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90239676_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092930_buildGRCh37.tsv.gz 16\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009249_buildGRCh37.txt 16\n",
      "Sumstat cols:\t['Variant,CHR,BP,ALLELE1,ALLELE0,A1FREQ,INFO,BETA,SE,P,id'] ->\n",
      "GPT mapping:\t['CHR', 'BP', 'ALLELE1', 'ALLELE0', 'BETA', 'SE', 'P', 'NA']\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093016_buildGRCh37.tsv.gz 16\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092868_buildGRCh37.tsv.gz 16\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269497_buildGRCh37.tsv.gz 16\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'info', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90269497_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006907_buildGRCh37.tsv 17\n",
      "Sumstat cols:\t['MarkerName Allele1 Allele2 Freq1 Effect StdErr P-value'] ->\n",
      "GPT mapping:\t['NA', 'NA', 'Allele1', 'Allele2', 'Effect', 'StdErr', 'P-value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90027266_buildGRCh37.tsv 17\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'variant_id', 'other_allele', 'effect_allele', 'MAF', 'imputationInfo', 'n', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90027266_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269616_buildGRCh37.tsv.gz 18\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'info', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90269616_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90274723_buildGRCh37.tsv 19\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'info', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90274723_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST011365_buildGRCh37.tsv 20\n",
      "Sumstat cols:\t['variant_id', 'p_value', 'chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'odds_ratio', 'ci_lower', 'ci_upper', 'beta', 'standard_error'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST011365_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093004_buildGRCh37.tsv.gz 21\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269904_buildGRCh37.tsv.gz 21\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90269904_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092944_buildGRCh37.tsv.gz 22\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009250_buildGRCh37.txt 22\n",
      "Sumstat cols:\t['Variant,CHR,BP,ALLELE1,ALLELE0,A1FREQ,INFO,BETA,SE,P,id'] ->\n",
      "GPT mapping:\t['CHR', 'BP', 'ALLELE1', 'ALLELE0', 'BETA', 'SE', 'P', 'NA']\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90086134_buildGRCh37.tsv 22\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'rs_id_all', 'info_all', 'other_allele', 'effect_allele', 'all_maf', 'all_total', 'freq_effect_allele', 'standard_error', 'frequentist_gen_se_2', 'beta', 'frequentist_gen_beta_2', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90086134_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90271714_buildGRCh37.tsv 23\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'N'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90271714_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092985_buildGRCh37.tsv.gz 24\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90269498_buildGRCh37.tsv.gz 24\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'info', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90269498_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90013537_buildGRCh37.tsv 25\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'other_allele', 'effect_allele', 'or.khor', 'se.khor', 'p.khor', 'score.new', 'info.new', 'p.new', 'z.meta', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'NA', 'se.khor', 'p.khor', 'or.khor']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90093028_buildGRCh37.tsv.gz 25\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092956_buildGRCh37.tsv.gz 25\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'NA', 'NA', 'allele2', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90310293_buildGRCh37.tsv.gz 25\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'variant_id', 'rs_id', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90310293_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90092968_buildGRCh37.tsv.gz 26\n",
      "Sumstat cols:\t['variant_id', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['variant_id', 'NA', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90239664_buildGRCh37.tsv 26\n",
      "Sumstat cols:\t['variant_id', 'chromosome', 'base_pair_location', 'other_allele', 'effect_allele', 'n', 'N_studies', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90239664_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90132315_buildGRCh37.tsv 27\n",
      "Sumstat cols:\t['p_value', 'chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'odds_ratio', 'beta', 'standard_error', 'markername', 'freqse', 'minfreq', 'maxfreq', 'direction', 'hetisq', 'hetchisq', 'hetdf', 'hetpval', 'cases', 'effective_cases', 'n'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST90132315_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90043977_buildGRCh37.tsv.gz 28\n",
      "Sumstat cols:\t['chromosome', 'variant_id', 'base_pair_location', 'effect_allele', 'other_allele', 'N', 'effect_allele_frequency', 'T', 'SE_T', 'P_noSPA', 'beta', 'standard_error', 'p_value', 'CONVERGE'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90043977_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90104542_buildGRCh37.tsv.gz 29\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele_frequency', 'beta', 'standard_error', 'p_value', 'odds_ratio', 'ci_lower', 'ci_upper', 'effect_allele', 'other_allele'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST90104542_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90271713_buildGRCh37.tsv 30\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'rs_id', 'N'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90271713_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST007546_buildGRCh37.txt 31\n",
      "Sumstat cols:\t['SNP Position EA OA beta.CAD beta.lipid eaf.CAD eaf.lipid se.lipid pval.lipid se.CAD pval.CAD P'] ->\n",
      "GPT mapping:\t['NA', 'Position', 'EA', 'OA', 'beta.CAD', 'se.CAD', 'pval.CAD', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST009602_buildGRCh37.txt 31\n",
      "Sumstat cols:\t['snpid', 'chr', 'pos', 'effect_allele', 'eaf', 'ntotal', 'beta', 'se', 'p_value', 'rsq', 'other_allele', 'rsid'] ->\n",
      "GPT mapping:\t['chr', 'pos', 'effect_allele', 'other_allele', 'beta', 'se', 'p_value', 'NA']\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006909_buildGRCh37.tsv 31\n",
      "Sumstat cols:\t['MarkerName Allele1 Allele2 Freq1 Effect StdErr P-value'] ->\n",
      "GPT mapping:\t['NA', 'NA', 'Allele1', 'Allele2', 'Effect', 'StdErr', 'P-value', 'NA']\n",
      "GPT couldn't make out all the columns.\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST004787_buildGRCh37.txt.gz 31\n",
      "Sumstat cols:\t['Markername', 'snptestid', 'chr', 'bp_hg19', 'effect_allele', 'noneffect_allele', 'effect_allele_freq', 'logOR', 'se_gc', 'p-value_gc', 'n_samples', 'exome', 'info_ukbb'] ->\n",
      "GPT mapping:\t['chr', 'bp_hg19', 'effect_allele', 'noneffect_allele', 'logOR', 'se_gc', 'p-value_gc', 'NA']\n",
      "SKIPPING due to error\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90301954_buildGRCh37.tsv 31\n",
      "Sumstat cols:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'effect_allele_frequency', 'p_value', 'variant_id', 'direction', 'hetisq', 'hetchisq', 'hetdf', 'hetpval'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90301954_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90044025_buildGRCh37.tsv.gz 32\n",
      "Sumstat cols:\t['chromosome', 'variant_id', 'base_pair_location', 'effect_allele', 'other_allele', 'N', 'effect_allele_frequency', 'T', 'SE_T', 'P_noSPA', 'beta', 'standard_error', 'p_value', 'CONVERGE'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'NA']\n",
      "Saved reformatted sumstats file to: GCST90044025_buildGRCh37_preprocessed.tsv\n",
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST006414_buildGRCh37.tbl.gz 33\n",
      "Sumstat cols:\t['MarkerName', 'rs_dbSNP147', 'CHR', 'POS_GRCh37', 'A1', 'A2', 'Freq_A2', 'Effect_A2', 'StdErr', 'Pvalue'] ->\n",
      "GPT mapping:\t['CHR', 'POS_GRCh37', 'A1', 'A2', 'Effect_A2', 'StdErr', 'Pvalue', 'NA']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/groups/sanjana_lab/sfriedman/finemapping/FiMTools/src/preprocessing_v04.py:266: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for chunk in pd.read_table(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved reformatted sumstats file to: GCST006414_buildGRCh37_preprocessed.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/groups/sanjana_lab/sfriedman/finemapping/FiMTools/src/preprocessing_v04.py:354: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for chunk in pd.read_table(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> /gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/raw/European/GCST90132314_buildGRCh37.tsv 34\n",
      "Sumstat cols:\t['p_value', 'chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'effect_allele_frequency', 'odds_ratio', 'beta', 'standard_error', 'markername', 'freqse', 'minfreq', 'maxfreq', 'direction', 'hetisq', 'hetchisq', 'hetdf', 'hetpval', 'cases', 'effective_cases', 'n', 'meta_analysis'] ->\n",
      "GPT mapping:\t['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'standard_error', 'p_value', 'odds_ratio']\n",
      "Saved reformatted sumstats file to: GCST90132314_buildGRCh37_preprocessed.tsv\n",
      "35\n"
     ]
    }
   ],
   "source": [
    "ct=0\n",
    "for path in my_input_folder[::]:\n",
    "\n",
    "    print(f'==> {path} {ct}')\n",
    "\n",
    "    try:\n",
    "\n",
    "        res = ft.loadmap_sumstats_table(\n",
    "            path,\n",
    "            verbose=False)\n",
    "        \n",
    "        ct += res\n",
    "\n",
    "        if res == 1:\n",
    "            ft.create_leadsnp_table(verbose=False)\n",
    "    except:\n",
    "        print(\"SKIPPING due to error\")\n",
    "     \n",
    "print(ct)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verify files were outputted to the specified directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90026612_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104543_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104542_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90245878_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90027266_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043954_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90239676_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90271714_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086137_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90271714_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086134_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086137_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90014122_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086134_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043977_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269905_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086092_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90132315_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90274723_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043954_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90132314_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90245878_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90301954_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90239664_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043977_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90044025_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104541_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90239664_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90239676_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043957_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269904_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST011365_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90274723_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90132314_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90044025_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST010774_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269498_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086077_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086077_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269904_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269905_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90310293_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90271713_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/column_mappings.log',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90301954_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST006414_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90025953_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90092831_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90027266_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086091_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90092831_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90271713_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90310293_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086091_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269616_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269616_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90025953_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90104542_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90043957_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90086092_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST011365_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269498_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269497_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90014122_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/prior',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90269497_buildGRCh37_preprocessed_leadSNPs.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST90132315_buildGRCh37_preprocessed.tsv',\n",
       " '/gpfs/commons/groups/sanjana_lab/mdrabkin/gwas_data/preprocessed/European/GCST006414_buildGRCh37_preprocessed_leadSNPs.tsv']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fps = glob.glob(f'{output_directory}/*')\n",
    "fps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
