---
title: "setting_up"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
setwd("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune")
library(data.table)
library(dplyr)
library(ggplot2)
source("src/finemapping_functions.R")
```

```{r}
gwasEUR_og <- fread("../AID_GWAS/SumStats/IBD/natgen_2015_PMID26192919/EUR.IBD.gwas_info03_filtered.assoc")
#gwasMULT <- fread("../AID_GWAS/SumStats/IBD/natgen_2015_PMID26192919/trans_ethnic_Immunochip/IBD_trans_ethnic_association_summ_stats_b37.txt")
leadSNPs <- fread("../AID_GWAS/SumStats/IBD/natgen_2015_PMID26192919/Lead_SNPS_IBD_NATGEN_2015.csv")
colnames(leadSNPs) <- c("chr", "og_SNP", "og_SNP_BP37", "pheno", "BF", "top_SNP", "top_SNP_BP37", "P", "R2")

```


```{r}

#df.leadSNP has columns BP, CHR, locus
#df.sumstats has columns chromosome, position, allele1, allele2, assuming allele1 is effect allele, beta, se
leadSNPs.1 <- leadSNPs %>% 
  filter(pheno == "IBD") %>% 
  mutate(locus = paste0(chr, ":", top_SNP_BP37)) %>% 
  select(CHR = chr, BP = top_SNP_BP37, locus)

gwasEUR <- gwasEUR_og %>%  
  mutate(beta = log(OR)) %>% 
  select(rsid = SNP, chromosome = CHR, position = BP, allele1 = A1, allele2 = A2, beta, se = SE) 
write_tsv(gwasEUR, "data/IBD.EUR.assoc")
```

```{r}
generate_per_locus_data_and_submit_pipeline(df.leadSNP = leadSNPs.1, "data/IBD.EUR.assoc", "EUR.IBD", 34694, 12924, 0.25, "EUR")
```

