---
title: "run_finemapping.Rmd"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
setwd("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune")
library(data.table)
library(dplyr)
library(ggplot2)
library(stringr)
library(tidyr)
library(readr)
source("src/finemapping_functions.R")
```


RA:
Read in GWAS and lead SNP info:
```{r}
gwasEUR_og <- fread("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/AID_GWAS/SumStats/RA/natgen_2022_PMID36333501/EUR_all_auto-10-2021.txt.gz")

leadSNPs <- fread("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/AID_GWAS/SumStats/RA/natgen_2022_PMID36333501/RA_Lead_Snps-EUR.csv")
colnames(leadSNPs) <- c("locus_ID", "rsid", "variant_ID", "chr", "position", "pop", "serostatus", "OR", "L95", "U95", "pval", "direction", "p.het")
```

prepare lead SNP df and GWAS, save prepared GWAS to file so that we can use SLURM to parallelize finemapping by locus
```{r}
#df.leadSNP has columns BP, CHR, locus
#df.sumstats has columns chromosome, position, allele1 (assuming allele1 is effect allele), allele2, beta, se
leadSNPs.1 <- leadSNPs %>% 
  mutate(position = as.numeric(str_replace_all(position, ",", "")))%>% 
  mutate(locus = paste0(chr, ":", position)) %>% 
  select(CHR = chr, BP = position, locus) 

gwasEUR <- gwasEUR_og %>%  
  separate(SNP, into = c("chromsome", "position", "A2", "A1"))
gwasEUR <- gwasEUR %>% 
  select(chromosome = chromsome, position, allele1 = A1, allele2 = A2, beta = Beta, se = SE) 
write_tsv(gwasEUR, "/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune/data/RA.EUR.assoc")
```

```{r}
#22,350 cases and 74823 controls
setwd("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune")
generate_per_locus_data_and_submit_pipeline(df.leadSNP = leadSNPs.1, 
                                            path_df_sumstats = "data/RA.EUR.assoc", 
                                            sumstats_name = "2022_RA_EUR", 
                                            N_tot = 22350 + 74823, 
                                            N_cases = 22350, 
                                            window_mb = 0.25, 
                                            ld_pop = "EUR")
```


T1D:
Read in GWAS and lead SNP info:
```{r}
gwasEUR_og <- fread("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/AID_GWAS/SumStats/T1D/nature_2021_PMID34012112/GCST90014023_buildGRCh37.tsv")

leadSNPs <- fread("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/AID_GWAS/SumStats/T1D/nature_2021_PMID34012112/LEAD_SNPs_T1D_Lead_SNP_Finemapped.csv")
colnames(leadSNPs) <- c("rsid", "chr", "position19", "position38", "ref", "alt", "signal", "alt_af", "PPA", "index_r2", "PMID")
```

prepare lead SNP df and GWAS, save prepared GWAS to file so that we can use SLURM to parallelize finemapping by locus
```{r}
#df.leadSNP has columns BP, CHR, locus
#df.sumstats has columns chromosome, position, allele1 (assuming allele1 is effect allele), allele2, beta, se
leadSNPs.1 <- leadSNPs %>% 
  mutate(locus = paste0(chr, ":", position19)) %>% 
  select(CHR = chr, BP = position19, locus) 

gwasEUR <- gwasEUR_og %>% 
  select(rsid = variant_id, chromosome = chr, position = pos, allele1 = effect_allele, allele2 = other_allele, beta, se = standard_error) 
write_tsv(gwasEUR, "/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune/data/T1D.EUR.assoc")
```

```{r}
#18,942 T1D cases and 501,638 controls
setwd("/gpfs/commons/groups/nygcfaculty/lappalainen_singh/finemapping_autoimmune")
generate_per_locus_data_and_submit_pipeline(df.leadSNP = leadSNPs.1, 
                                            path_df_sumstats = "data/T1D.EUR.assoc", 
                                            sumstats_name = "2021_T1D_EUR", 
                                            N_tot = 501638 + 18942, 
                                            N_cases = 18942, 
                                            window_mb = 0.25, 
                                            ld_pop = "EUR")
```